src/main/java/net/neoforged/neoforge/transfer/handlers/wrappers/items/ItemContainerContentsResourceHandler.java

```diff
@@ -0,0 +1,146 @@
+        ItemContainerContents contents = getContents();
+        return getStackInSlot(contents, index).getCount();
+    }
+
+    @Override
+    public int getCapacity(int index, ItemResource resource) {
+        Objects.checkIndex(index, size());
+        if (resource.isEmpty()) return Item.ABSOLUTE_MAX_STACK_SIZE;
+        return Math.min(resource.getMaxStackSize(), Item.ABSOLUTE_MAX_STACK_SIZE);
+    }
+
+    @Override
+    public boolean isValid(int index, ItemResource resource) {
+        Objects.checkIndex(index, size());
+        return resource.getInstanceValue().canFitInsideContainerItems();
+    }
+
+    @Override
+    public boolean supportsInsertion(int index) {
+        Objects.checkIndex(index, size());
+        return true;
+    }
+
+    @Override
+    public boolean supportsExtraction(int index) {
+        Objects.checkIndex(index, size());
+        return true;
+    }
+
+    private ItemStack getStackInSlot(ItemContainerContents contents, int index) {
+        //Index bounds has already been checked by this point
+        return contents.getSlots() <= index ? ItemStack.EMPTY : contents.getStackInSlot(index);
+    }
+
+    @Override
+    public int insert(int index, ItemResource resource, int amount, TransactionContext context) {
+        Objects.checkIndex(index, size());
+        if (ResourceHandlerUtil.isEmpty(resource, amount) || !isValid(index, resource)) return 0;
+
+        ItemContainerContents contents = getContents();
+        ItemStack stack = getStackInSlot(contents, index);
+
+        if (stack.isEmpty()) {
+            int inserted = Math.min(amount, resource.getMaxStackSize());
+            return set(contents, inserted, context, index, resource.toStack(inserted));
+        }
+
+        if (!resource.is(stack) || stack.getCount() >= resource.getMaxStackSize()) return 0;
+
+        int inserted = Math.min(amount, getCapacity(index, resource) - stack.getCount());
+        stack.grow(inserted);
+        return set(contents, inserted, context, index, stack);
+    }
+
+    @Override
+    public int extract(int index, ItemResource resource, int amount, TransactionContext context) {
+        Objects.checkIndex(index, size());
+        if (ResourceHandlerUtil.isEmpty(resource, amount)) return 0;
+
+        ItemContainerContents contents = getContents();
+        ItemStack stack = getStackInSlot(contents, index);
+
+        if (stack.isEmpty() || !resource.is(stack)) return 0;
+
+        int extracted = Math.min(stack.getCount(), amount);
+        stack.shrink(extracted);
+        return set(contents, amount, context, index, stack);
+    }
+
+    /**
+     * Handles changing the component out on the context container
+     */
+    @ApiStatus.OverrideOnly
+    protected int set(ItemContainerContents contents, int changedAmount, TransactionContext context, int index, ItemStack stack) {
+        ItemResource contextResource = itemContext.getResource();
+        ItemStack newStack = contextResource.toStack();
```

Comment:
Why make this stack instead of just using the helpers for adjusting the components that exist on `ItemResource`?